examples/basic_authentication/lib/user_repository.dart

```diff
@@ -0,0 +1,120 @@
+import 'dart:convert';
+
+import 'package:crypto/crypto.dart';
+import 'package:equatable/equatable.dart';
+import 'package:meta/meta.dart';
+
+@visibleForTesting
+
+/// In memory database of users.
+Map<String, User> db = {};
+
+/// {@template user}
+/// A User.
+/// {@endtemplate}
+class User extends Equatable {
+  /// {@macro user}
+  const User({
+    required this.id,
+    required this.name,
+    required this.username,
+    required this.password,
+  });
+
+  /// The user's id.
+  final String id;
+
+  /// The user's name.
+  final String name;
+
+  /// The user's username.
+  final String username;
+
+  /// The user's password, in a hashed form.
+  final String password;
+
+  @override
+  List<Object?> get props => [id, name, username, password];
```

Comment:
Nit: to consider two users to be the same only if they share the same id?

```suggestion
  List<Object?> get props => [id];
```

Reply:
In theory, yes, but we also need to take in account all of the fields, because imagine in a `updateUser` method, we will change some fields of the model given the new input data from the user. Then the id will be the same, but those instances are not equal anymore, even though both instances point to the same user.